tests/test_other.py

```diff
@@ -3634,25 +3721,41 @@ def test_bad_function_pointer_cast(self):
         for emulate_casts in [0, 1]:
           for emulate_fps in [0, 1]:
             for relocate in [0, 1]:
-              cmd = [PYTHON, EMCC, 'src.cpp', '-O' + str(opts), '-s', 'SAFE_HEAP=' + str(safe)]
-              if emulate_casts:
-                cmd += ['-s', 'EMULATE_FUNCTION_POINTER_CASTS=1']
-              if emulate_fps:
-                cmd += ['-s', 'EMULATED_FUNCTION_POINTERS=1']
-              if relocate:
-                cmd += ['-s', 'RELOCATABLE=1'] # disables asm-optimized safe heap
-              print(cmd)
-              Popen(cmd).communicate()
-              output = run_js('a.out.js', stderr=PIPE, full_output=True, assert_returncode=None)
-              if emulate_casts:
-                assert 'Hello, world.' in output, output
-              elif safe:
-                assert 'Function table mask error' in output, output
-              else:
-                if opts == 0:
-                  assert 'Invalid function pointer called' in output, output
+              for wasm in [0, 1]:
+                cmd = [PYTHON, EMCC, 'src.cpp', '-O' + str(opts), '-s', 'SAFE_HEAP=' + str(safe), '-s', 'WASM=' + str(wasm)]
+                if emulate_casts:
+                  cmd += ['-s', 'EMULATE_FUNCTION_POINTER_CASTS=1']
+                if emulate_fps:
+                  cmd += ['-s', 'EMULATED_FUNCTION_POINTERS=1']
+                if relocate:
+                  cmd += ['-s', 'RELOCATABLE=1'] # disables asm-optimized safe heap
+                print(cmd)
+                Popen(cmd).communicate()
+                output = run_js('a.out.js', stderr=PIPE, full_output=True, assert_returncode=None)
+                if emulate_casts:
+                  assert 'Hello, world.' in output, output
+                elif safe:
+                  if wasm:
+                    if relocate or emulate_fps:
+                      assert 'function signature mismatch' in output, output
+                    else:
+                      if opts == 0:
+                        assert 'Invalid function pointer called' in output, output
+                      else:
+                        assert 'abort(' in output, output
+                  else:
+                    assert 'Function table mask error' in output, output
                 else:
-                  assert 'abort(' in output, output
+                  if opts == 0:
+                    if wasm and (relocate or emulate_fps):
+                      assert 'function signature mismatch' in output, output
```

Comment:
This is all a tad horrifying
Maybe something like:
```python
for emulate_casts, emulate_fps, relocate, wasm, expected in [
  (0, 0, 0, 0, 'Invalid function pointer called'),
  (0, 0, 0, 1, 'Invalid function pointer called'),
  ...
]:
  ...
  assert expected in output, 'Expected to see ' + expected + ' in ' + output
```
Generally flatten it out and make it clearer what preconditions connect with what output

Reply:
Hmm, there are 96 combinations, though, so flattening it all out seems unpleasant. But I'll try to flatten out those if-else chains, those do seem ugly I agree.

Reply:
Flattened out the ifs and added clarifying comments.

Reply:
Oh I missed the outer two loops, I thought it was only 16 total which seemed less bad. In theory one could also only test a subset to reduce the cases further, though that trades off against completeness.

Regardless, the flattening (and clarifying comments) works even better, so let's keep that.